File: app/src/main/java/mega/privacy/android/app/repo/MegaNodeRepo.java
package mega.privacy.android.app.repo;

import android.content.Context;
import android.text.TextUtils;
import android.util.Pair;
import java.time.LocalDate;
import java.time.YearMonth;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.Function;
import mega.privacy.android.app.DatabaseHandler;
import mega.privacy.android.app.MegaPreferences;
import mega.privacy.android.app.MimeTypeThumbnail;
import mega.privacy.android.app.R;
import nz.mega.sdk.MegaApiAndroid;
import nz.mega.sdk.MegaNode;

import static mega.privacy.android.app.utils.LogUtil.logError;
import static mega.privacy.android.app.utils.Util.fromEpoch;

public class MegaNodeRepo {

    public static final int CU_TYPE_CAMERA = 0;
    public static final int CU_TYPE_MEDIA = 1;

    private final MegaApiAndroid megaApi;
    private final DatabaseHandler dbHandler;
    private final Context context;

    public MegaNodeRepo(MegaApiAndroid megaApi, DatabaseHandler dbHandler, Context context) {
        this.megaApi = megaApi;
        this.dbHandler = dbHandler;
        this.context = context;
    }

    /**
     * Get children of CU/MU, with the given order, and filter nodes by date (optional).
     *
     * @param type CU_TYPE_CAMERA or CU_TYPE_MEDIA
     * @param orderBy order
     * @param filter search filter
     * filter[0] is the search type:
     * 0 means search for nodes in one day, then filter[1] is the day in millis.
     * 1 means search for nodes in last month (filter[2] is 1), or in last year (filter[2] is 2).
     * 2 means search for nodes between two days, filter[3] and filter[4] are start and end day in
     * millis.
     * @return list of pairs, whose first value is index in parent, second value is the node, and
     * the index of parent is used for FullscreenImageViewer/AudioVideoPlayer
     */
    public List<Pair<Integer, MegaNode>> getCuChildren(int type, int orderBy, long[] filter) {
        long cuHandle = -1;
        MegaPreferences pref = dbHandler.getPreferences();
        if (type == CU_TYPE_CAMERA) {
            if (pref != null && pref.getCamSyncHandle() != null) {
                try {
                    cuHandle = Long.parseLong(pref.getCamSyncHandle());
                } catch (NumberFormatException e) {
                    logError("parse getCamSyncHandle error " + e);
                }
                if (megaApi.getNodeByHandle(cuHandle) == null) {
                    cuHandle = -1;
                }
            }

            if (cuHandle == -1) {
                for (MegaNode node : megaApi.getChildren(megaApi.getRootNode())) {
                    if (node.isFolder() && TextUtils.equals(
                            context.getString(R.string.section_photo_sync),
                            node.getName())) {
                        cuHandle = node.getHandle();
                        dbHandler.setCamSyncHandle(cuHandle);
                        break;
                    }
                }
            }
        } else {
            if (pref != null && pref.getMegaHandleSecondaryFolder() != null) {
                try {
                    cuHandle = Long.parseLong(pref.getMegaHandleSecondaryFolder());
                } catch (NumberFormatException e) {
                    logError("parse MegaHandleSecondaryFolder error " + e);
                }
                if (megaApi.getNodeByHandle(cuHandle) == null) {
                    cuHandle = -1;
                }
            }
        }

        if (cuHandle == -1) {
            return Collections.emptyList();
        }

        List<MegaNode> children = megaApi.getChildren(megaApi.getNodeByHandle(cuHandle), orderBy);
        List<Pair<Integer, MegaNode>> nodes = new ArrayList<>();
        for (int i = 0; i < children.size(); i++) {
            MegaNode node = children.get(i);
            if (node.isFolder()) {
                continue;
            }
            MimeTypeThumbnail mime = MimeTypeThumbnail.typeForName(node.getName());
            if (mime.isImage() || mime.isVideoReproducible()) {
                nodes.add(Pair.create(i, node));
            }
        }

        if (filter == null) {
            return nodes;
        }

        List<Pair<Integer, MegaNode>> result = new ArrayList<>();

        Function<MegaNode, Boolean> filterFunction = null;
        if (filter[0] == 1) {
            LocalDate date = fromEpoch(filter[1] / 1000);
            filterFunction = node -> date.equals(fromEpoch(node.getModificationTime()));
        } else if (filter[0] == 2) {
            if (filter[2] == 1) {
                YearMonth lastMonth = YearMonth.now().minusMonths(1);
                filterFunction =
                        node -> lastMonth.equals(
                                YearMonth.from(fromEpoch(node.getModificationTime())));
            } else if (filter[2] == 2) {
                int lastYear = YearMonth.now().getYear() - 1;
                filterFunction =
                        node -> fromEpoch(node.getModificationTime()).getYear() == lastYear;
            }
        } else if (filter[0] == 3) {
            LocalDate from = fromEpoch(filter[3] / 1000);
            LocalDate to = fromEpoch(filter[4] / 1000);
            filterFunction = node -> {
                LocalDate modifyDate = fromEpoch(node.getModificationTime());
                return !modifyDate.isBefore(from) && !modifyDate.isAfter(to);
            };
        }

        if (filterFunction == null) {
            return result;
        }

        for (Pair<Integer, MegaNode> pair : nodes) {
            if (filterFunction.apply(pair.second)) {
                result.add(pair);
            }
        }

        return result;
    }
}

File: app/src/main/java/mega/privacy/android/app/repo/MegaNodeRepo.kt
package mega.privacy.android.app.repo

import mega.privacy.android.app.LegacyDatabaseHandler
import mega.privacy.android.app.MegaApplication
import mega.privacy.android.app.MegaOffline
import mega.privacy.android.app.MimeTypeList
import mega.privacy.android.app.utils.FileUtil
import mega.privacy.android.app.utils.OfflineUtils.getOfflineFile
import mega.privacy.android.app.utils.SortUtil.sortOfflineByModificationDateAscending
import mega.privacy.android.app.utils.SortUtil.sortOfflineByModificationDateDescending
import mega.privacy.android.app.utils.SortUtil.sortOfflineByNameAscending
import mega.privacy.android.app.utils.SortUtil.sortOfflineByNameDescending
import mega.privacy.android.app.utils.SortUtil.sortOfflineBySizeAscending
import mega.privacy.android.app.utils.SortUtil.sortOfflineBySizeDescending
import mega.privacy.android.data.qualifier.MegaApi
import mega.privacy.android.domain.entity.SortOrder
import nz.mega.sdk.MegaApiAndroid
import nz.mega.sdk.MegaApiJava.ORDER_DEFAULT_ASC
import nz.mega.sdk.MegaApiJava.ORDER_DEFAULT_DESC
import nz.mega.sdk.MegaApiJava.ORDER_MODIFICATION_ASC
import nz.mega.sdk.MegaApiJava.ORDER_MODIFICATION_DESC
import nz.mega.sdk.MegaApiJava.ORDER_SIZE_ASC
import nz.mega.sdk.MegaApiJava.ORDER_SIZE_DESC
import nz.mega.sdk.MegaNode
import nz.mega.sdk.MegaNodeList
import timber.log.Timber
import java.io.File
import java.util.Locale
import javax.inject.Inject

class MegaNodeRepo @Inject constructor(
    @MegaApi private val megaApi: MegaApiAndroid,
    private val dbHandler: LegacyDatabaseHandler
) {

    /**
     * Get children of CU/MU, with the given order. All nodes, including folders.
     *
     * @param orderBy Order.
     * @return List of nodes containing CU/MU children.
     */
    private fun getCuChildren(orderBy: Int): List<MegaNode> {
        var cuNode: MegaNode? = null
        var muNode: MegaNode? = null
        val pref = dbHandler.preferences

        if (pref?.camSyncHandle != null) {
            try {
                val cuHandle = pref.camSyncHandle.toLong()
                cuNode = megaApi.getNodeByHandle(cuHandle)
            } catch (e: NumberFormatException) {
                Timber.e(e, "parse getCamSyncHandle error")
            }
        }

        if (pref?.megaHandleSecondaryFolder != null) {
            try {
                val muHandle = pref.megaHandleSecondaryFolder.toLong()
                muNode = megaApi.getNodeByHandle(muHandle)
            } catch (e: NumberFormatException) {
                Timber.e(e, "parse MegaHandleSecondaryFolder error")
            }
        }

        if (cuNode == null && muNode == null) {
            return emptyList()
        }

        val nodeList = MegaNodeList.createInstance()

        if (cuNode != null) {
            nodeList.addNode(cuNode)
        }

        if (muNode != null) {
            nodeList.addNode(muNode)
        }

        return megaApi.getChildren(nodeList, orderBy)
    }

    /**
     * Get children of CU/MU, with the given order. Only images and reproducible videos.
     *
     * @param orderBy Order.
     * @return List of nodes containing CU/MU children.
     */
    fun getFilteredCuChildren(
        orderBy: Int
    ): List<MegaNode> {
        val children = getCuChildren(orderBy)
        val nodes = ArrayList<MegaNode>()

        for (node in children) {
            if (megaApi.isInRubbish(node))
                continue

            if (node.isFolder) {
                continue
            }

            val mime = MimeTypeList.typeForName(node.name)
            if (mime.isImage || mime.isVideoReproducible) {
                nodes.add(node)
            }
        }

        return nodes
    }

    fun findOfflineNode(handle: String): MegaOffline? {
        return dbHandler.findByHandle(handle)
    }

    fun loadOfflineNodes(path: String, order: SortOrder, searchQuery: String?): List<MegaOffline> {
        val nodes = if (searchQuery != null && searchQuery.isNotEmpty()) {
            searchOfflineNodes(path, searchQuery)
        } else {
            dbHandler.findByPath(path)
        }

        when (order) {
            SortOrder.ORDER_DEFAULT_DESC -> {
                sortOfflineByNameDescending(nodes)
            }
            SortOrder.ORDER_DEFAULT_ASC -> {
                sortOfflineByNameAscending(nodes)
            }
            SortOrder.ORDER_MODIFICATION_ASC -> {
                sortOfflineByModificationDateAscending(nodes)
            }
            SortOrder.ORDER_MODIFICATION_DESC -> {
                sortOfflineByModificationDateDescending(nodes)
            }
            SortOrder.ORDER_SIZE_ASC -> {
                sortOfflineBySizeAscending(nodes)
            }
            SortOrder.ORDER_SIZE_DESC -> {
                sortOfflineBySizeDescending(nodes)
            }
            else -> {
            }
        }
        return nodes
    }

    private fun searchOfflineNodes(path: String, query: String): ArrayList<MegaOffline> {
        val result = ArrayList<MegaOffline>()

        val nodes = dbHandler.findByPath(path)
        for (node in nodes) {
            if (node.isFolder) {
                result.addAll(searchOfflineNodes(getChildPath(node), query))
            }

            if (node.name.lowercase(Locale.ROOT).contains(query.lowercase(Locale.ROOT)) &&
                FileUtil.isFileAvailable(getOfflineFile(MegaApplication.getInstance(), node))
            ) {
                result.add(node)
            }
        }

        return result
    }

    private fun getChildPath(offline: MegaOffline): String {
        return if (offline.path.endsWith(File.separator)) {
            offline.path + offline.name + File.separator
        } else {
            offline.path + File.separator + offline.name + File.separator
        }
    }
}
